src/boot_O2/gfxprint.c

```diff
@@ -93,49 +86,95 @@ void GfxPrint_PrintCharImpl(GfxPrint* this, u8 c) {
         }
     }
 
-    if (this->flag & GFXPRINT_FLAG4) {
-        gDPSetPrimColorMod(this->dlist++, 0, 0, 0);
+    if (this->flags & GFXP_FLAG_SHADOW) {
+        gDPSetColor(this->dlist++, G_SETPRIMCOLOR, 0);
 
         gSPTextureRectangle(this->dlist++, this->posX + 4, this->posY + 4, this->posX + 4 + 32, this->posY + 4 + 32,
-                            (c & 3) << 1, (u16)(c & 4) * 64, (u16)(c >> 3) * 256, 1024, 1024);
+                            tile, s << 6, t << 8, 1 << 10, 1 << 10);
 
-        gDPSetPrimColorMod(this->dlist++, 0, 0, this->color.rgba);
+        gDPSetColor(this->dlist++, G_SETPRIMCOLOR, this->color.rgba);
     }
 
-    gSPTextureRectangle(this->dlist++, this->posX, this->posY, this->posX + 32, this->posY + 32, (u16)(tile & 7),
-                        (u16)(c & 4) * 64, (u16)(c >> 3) * 256, 1024, 1024);
+    gSPTextureRectangle(this->dlist++, this->posX, this->posY, this->posX + 32, this->posY + 32, tile, s << 6, t << 8,
+                        1 << 10, 1 << 10);
 
     this->posX += 32;
 }
-#else
-#pragma GLOBAL_ASM("asm/non_matchings/boot/gfxprint/GfxPrint_PrintCharImpl.s")
-#endif
 
-#pragma GLOBAL_ASM("asm/non_matchings/boot/gfxprint/GfxPrint_PrintChar.s")
+void GfxPrint_PrintChar(GfxPrint* this, u8 c) {
+    if (c == ' ') {
+        this->posX += 32;
+    } else if (c > ' ' && c < 0x7F) {
+        GfxPrint_PrintCharImpl(this, c);
+    } else if (c >= 0xA0 && c < 0xE0) {
+        if (this->flags & GFXP_FLAG_HIRAGANA) {
+            if (c < 0xC0) {
+                c -= 0x20;
+            } else {
+                c += 0x20;
+            }
+        }
+        GfxPrint_PrintCharImpl(this, c);
+    } else {
+        switch (c) {
+            case '\0':
+                break;
+            case '\n':
+                this->posY += 32;
+            case '\r':
+                this->posX = this->baseX;
+                break;
+            case '\t':
+                do {
+                    GfxPrint_PrintCharImpl(this, ' ');
+                } while ((this->posX - this->baseX) % 256);
```

Comment:
I feel like some of these constants might be better in hex, although it depends how we want to do them across the board.

Reply:
Not sure, they might be better as defines. Probably best to wait for further docs.

Reply:
That's fair enough, just mentioning it in case other people want to disagree now.